File: LeetCode/897.cpp
//Link: https://leetcode.com/problems/increasing-order-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* head, *ptr;
    TreeNode* increasingBST(TreeNode* root) {
        create(root);
        return head;
    }
private:
    void create(TreeNode* node) {
        if(node == nullptr) return ;
        create(node->left);
        TreeNode* newNode = new TreeNode(node->val);
        if(head == nullptr) {
            head = newNode;
            ptr = newNode;
        }
        else {
            ptr->right = newNode;
            ptr = ptr->right;
        }
        create(node->right);
    }
};

File: LeetCode/897.py
#Link: https://leetcode.com/problems/increasing-order-search-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    head = ptr = None
    def increasingBST(self, root: TreeNode) -> TreeNode:
        self.create(root)
        return self.head
    def create(self, node):
        if node == None: return
        self.create(node.left)
        newNode = TreeNode(node.val)
        if self.head == None:
            self.head = newNode
            self.ptr = newNode
        else:
            self.ptr.right = newNode
            self.ptr = self.ptr.right
        self.create(node.right)
        